transports/janus_pfunix.c

```diff
@@ -0,0 +1,507 @@
+		g_thread_join(pfunix_thread);
+		pfunix_thread = NULL;
+	}
+
+	g_atomic_int_set(&initialized, 0);
+	g_atomic_int_set(&stopping, 0);
+	JANUS_LOG(LOG_INFO, "%s destroyed!\n", JANUS_PFUNIX_NAME);
+}
+
+int janus_pfunix_get_api_compatibility(void) {
+	/* Important! This is what your plugin MUST always return: don't lie here or bad things will happen */
+	return JANUS_TRANSPORT_API_VERSION;
+}
+
+int janus_pfunix_get_version(void) {
+	return JANUS_PFUNIX_VERSION;
+}
+
+const char *janus_pfunix_get_version_string(void) {
+	return JANUS_PFUNIX_VERSION_STRING;
+}
+
+const char *janus_pfunix_get_description(void) {
+	return JANUS_PFUNIX_DESCRIPTION;
+}
+
+const char *janus_pfunix_get_name(void) {
+	return JANUS_PFUNIX_NAME;
+}
+
+const char *janus_pfunix_get_author(void) {
+	return JANUS_PFUNIX_AUTHOR;
+}
+
+const char *janus_pfunix_get_package(void) {
+	return JANUS_PFUNIX_PACKAGE;
+}
+
+gboolean janus_pfunix_is_janus_api_enabled(void) {
+	return pfd > -1;
+}
+
+gboolean janus_pfunix_is_admin_api_enabled(void) {
+	return admin_pfd > -1;
+}
+
+int janus_pfunix_send_message(void *transport, void *request_id, gboolean admin, json_t *message) {
+	if(message == NULL)
+		return -1;
+	if(transport == NULL) {
+		g_free(message);
+		return -1;
+	}
+	/* Make sure this is related to a still valid Unix Sockets session */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) == NULL || client->fd < 0) {
+		janus_mutex_unlock(&clients_mutex);
+		JANUS_LOG(LOG_WARN, "Outgoing message for invalid client %p\n", client);
+		g_free(message);
+		message = NULL;
+		return -1;
+	}
+	janus_mutex_unlock(&clients_mutex);
+	/* Convert to string and enqueue */
+	char *payload = json_dumps(message, JSON_INDENT(3) | JSON_PRESERVE_ORDER);
+	json_decref(message);
+	g_async_queue_push(client->messages, payload);
+	/* Notify the thread there's data to send */
+	int res = 0;
+	do {
+		res = write(write_fd[1], "x", 1);
+	} while(res == -1 && errno == EINTR);
+	return 0;
+}
+
+void janus_pfunix_session_created(void *transport, guint64 session_id) {
+	/* We don't care */
+}
+
+void janus_pfunix_session_over(void *transport, guint64 session_id, gboolean timeout) {
+	/* We only care if it's a timeout: if so, close the connection */
+	if(transport == NULL || !timeout)
+		return;
+	/* FIXME Should we really close the connection in case of a timeout? */
+	janus_pfunix_client *client = (janus_pfunix_client *)transport;
+	janus_mutex_lock(&clients_mutex);
+	if(g_hash_table_lookup(clients, client) != NULL && client->fd > -1) {
+		/* Shutdown the client socket */
+		client->session_timeout = TRUE;
+		shutdown(client->fd, SHUT_WR);
+	}
+	janus_mutex_unlock(&clients_mutex);
+}
+
+
+/* Thread */
+void *janus_pfunix_thread(void *data) {
+	JANUS_LOG(LOG_INFO, "Unix Sockets thread started\n");
+
+	int fds = 0;
+	struct pollfd poll_fds[1024];	/* FIXME Should we allow for more clients? */
+	char buffer[BUFFER_SIZE];
```

Comment:
Might buffer size be a candidate for config? The difficulty with SOCK_SEQPACKET here is that there is no fixed max request size.